File: Source/GITTreeItem.m
//
//  GITTreeItem.m
//  Git.framework
//
//  Created by Geoff Garside on 14/12/2009.
//  Copyright 2009 Geoff Garside. All rights reserved.
//

#import "GITTreeItem.h"
#import "GITObjectHash.h"
#import "GITTree.h"
#import "GITRepo.h"


@implementation GITTreeItem

@synthesize parent, mode, name, item, sha1;

+ (GITTreeItem *)itemInTree: (GITTree *)tree withMode: (NSUInteger)mode name: (NSString *)name sha1: (GITObjectHash *)sha1 {
    return [[[self alloc] initInTree:tree withMode:mode name:name sha1:sha1] autorelease];
}

- (id)initInTree: (GITTree *)tree withMode: (NSUInteger)theMode name: (NSString *)theName sha1: (GITObjectHash *)theSha1 {
    if ( ![super init] )
        return nil;

    self.parent = tree;
    self.mode = theMode;
    self.name = theName;
    self.sha1 = theSha1;

    return self;
}

- (GITObject *)item {
    if ( !item )
        self.item = [self.parent.repo objectWithSha1:self.sha1 error:NULL];
    return item;
}

- (void)dealloc {
    self.parent = nil;
    self.name   = nil;
    self.sha1   = nil;
    if ( item )
        self.item = nil;

    [super dealloc];
}

- (BOOL)isLink {
    return (mode & GITTreeItemModeLink) == GITTreeItemModeLink;
}

- (BOOL)isFile {
    return (mode & GITTreeItemModeFile) == GITTreeItemModeFile;
}

- (BOOL)isDirectory {
    return (mode & GITTreeItemModeDir) == GITTreeItemModeDir;
}

- (BOOL)isModule {
    return (mode & GITTreeItemModeMod) == GITTreeItemModeMod;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"<GITTreeItem: %p sha1=%@, mode=%06x, name=%@>", self, self.sha1, self.mode, self.name];
}

- (NSUInteger)hash {
    return [self.sha1 hash];
}

- (BOOL)isEqual: (id)other {
    if ( !other ) return NO;                            // other is nil
    if ( other == self ) return YES;                    // pointers match?

    if ( [other isKindOfClass:[self class]] )           // Same class?
        return [self isEqualToTreeItem:other];
    if ( [other isKindOfClass:[GITObject class]] )      // A git object?
        return [self isEqualToObject:other];

    return NO;                                          // Definitely not then
}

- (BOOL)isEqualToTreeItem: (GITTreeItem *)rhs {
    if ( !rhs )         return NO;
    if ( self == rhs )  return YES;
    if ( [self.sha1 isEqualToObjectHash:[rhs sha1]] )
        return YES;
    return NO;
}

- (BOOL)isEqualToObject: (GITObject *)rhs {
    if ( !rhs ) return NO;
    if ( [self.sha1 isEqualToObjectHash:[rhs sha1]] )
        return YES;
    return NO;
}

@end

File: Test/Unit/GITTreeItem.rb
describe 'GITTreeItem' do
  before do
    @sha1 = GITObjectHash.objectHashWithString('afc72b2679ebc4d72b08ffe66d1adfb556d96f79')
    @tree = GITTree.alloc.init
    @treeItem = GITTreeItem.itemInTree(@tree, withMode:0x100644, name:".gitignore", sha1:@sha1)
  end
  describe '-hash' do
    should 'be equal to @sha1.hash' do
      @treeItem.hash.should == @sha1.hash
    end
  end
  describe '-isEqual:' do
    before do
      err = Pointer.new(:object)
      repo = simple_repository.git_repo
      @sha2 = GITObjectHash.objectHashWithString('a2d0bba4b0d4f9bf0168dd68d1588bde05aa3e5b')
      @blob = repo.objectWithSha1(@sha2, error:err)

      @treeItem2 = GITTreeItem.itemInTree(@tree, withMode:0x100644, name:"file.txt", sha1:@sha2)
    end
    should 'be equal to itself' do
      @treeItem.isEqual(@treeItem).should.be.true
    end
    should 'not be equal to object with different sha1' do
      @treeItem.isEqual(@treeItem2).should.be.false
    end
    should 'be equal to object with matching sha1' do
      @treeItem2.isEqual(@blob).should.be.true
    end
  end
end
